core/jvm/src/main/scala/scalaz/zio/lockfree/RingBuffer.scala

```diff
@@ -0,0 +1,147 @@
+package scalaz.zio.lockfree
+
+import java.util.concurrent.atomic.{ AtomicLong, AtomicReferenceArray }
+
+class RingBuffer[A](val desiredCapacity: Int) extends LockFreeQueue[A] {
+  final val capacity: Int                              = nextPow2(desiredCapacity)
+  private val buf: AtomicReferenceArray[BufferElement] = new AtomicReferenceArray[BufferElement](capacity)
+  private val idxMask: Long                            = (capacity - 1).toLong
+  private val head: AtomicLong                         = new AtomicLong(0L)
+  private val tail: AtomicLong                         = new AtomicLong(0L)
```

Comment:
Instead of having a `head` and `tail` `AtomicLong` (which means you can't change them both at the same time), I suggest a single `AtomicLong` that encodes _both_ `head` _and_ `tail`. We use the upper 32 bits for the head, and the lower 32 bits for the tail. Now we can atomically update them both in a single CAS.

Reply:
I'm not sure in what scenario we would need to change both head an tail at the same time. If we put head and tail in a single AtomicLong then we'll be hitting the same path during either offer or poll; while with separate atomics offer and poll can proceed independently without interfering with each other.